src/Query/QueryAccessHandlerBase.php

```diff
@@ -0,0 +1,210 @@
+<?php
+
+namespace Drupal\entity\Query;
+
+use Drupal\Core\Entity\EntityHandlerInterface;
+use Drupal\Core\Entity\EntityPublishedInterface;
+use Drupal\Core\Entity\EntityTypeBundleInfoInterface;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\user\EntityOwnerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides common logic for query access handlers.
+ *
+ * @see \Drupal\entity\Query\QueryAccessHandler
+ * @see \Drupal\entity\Query\UncacheableQueryAccessHandler
+ */
+abstract class QueryAccessHandlerBase implements EntityHandlerInterface, QueryAccessHandlerInterface {
+
+  /**
+   * The entity type.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeInterface
+   */
+  protected $entityType;
+
+  /**
+   * The entity type bundle info.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface
+   */
+  protected $bundleInfo;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $currentUser;
+
+  /**
+   * Constructs a new UncacheableQueryAccessHandler object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   *   The entity type.
+   * @param \Drupal\Core\Entity\EntityTypeBundleInfoInterface $bundle_info
+   *   The entity type bundle info.
+   * @param \Drupal\Core\Session\AccountInterface $current_user
+   *   The current user.
+   */
+  public function __construct(EntityTypeInterface $entity_type, EntityTypeBundleInfoInterface $bundle_info, AccountInterface $current_user) {
+    $this->entityType = $entity_type;
+    $this->bundleInfo = $bundle_info;
+    $this->currentUser = $current_user;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type) {
+    return new static(
+      $entity_type,
+      $container->get('entity_type.bundle.info'),
+      $container->get('current_user')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConditions($operation, AccountInterface $account = NULL) {
+    $account = $account ?: $this->currentUser;
+    $entity_type_id = $this->entityType->id();
+
+    if ($account->hasPermission("administer {$entity_type_id}")) {
+      // The user has full access to all operations, no conditions needed.
+      $conditions = new ConditionGroup('OR');
+      $conditions->addCacheContexts(['user.permissions']);
```

Comment:
@wimleers this is why `RefineableCacheableDependencyInterface` was needed.

@bojanz: In JSON API, we had similar value objects and dealt quite a bit with access related caching. We found it to be a major pain point and source of bugs. To help make caching more reasonable and less error-prone, @wimleers added a constructor argument to all the value objects and [got rid of the `RefineableCacheableDependency` stuff.](https://www.drupal.org/project/jsonapi/issues/2940342#comment-12517745). That essentially made everything immutable.

I recommend that cacheability be an explicit argument to the condition constructor. Given that we're dealing with access caching, I think it's acceptable, even advisable, to force implementors to think about caching, despite the slight DX loss and tediousness of it.

That means: removing `RefinableCacheableDependencyTrait`. Adding an argument to `ConditionGroup`'s constructor.

Also, why not make the Condition themselves cacheable, for most cases, they're really what are likely to change, not the group.

Reply:
This was attempted in https://github.com/bojanz/entity/pull/5, but it feels like the price in terms of DX regressions is too steep.

Reply:
I think two things are getting mixed together:

1. Making CacheableMetadata a required argument to `Condition::__construct`
2. Passing a `Condition` object to `ConditionGroup::addCondition` vs. a series of arguments and letting `addCondition` make the call to `new`.

I think your DX concern is actually more about 2 than 1?

That was honestly just a personal preference of mine that I should have kept out of the PR.

Making all arguments to `Condition::__construct` required, cacheability included, is the real change there and I think it's a worthy price to pay for correctness.

Reply:
https://github.com/fago/entity/pull/52#discussion_r199596243 already pointed out that `$operator` should probably just be required anyway.